### PR TITLE
feat: Add Damped Shifted Force (DSF) electrostatics

### PR DESCRIPTION
**Pull request by zubatyuk**
## Description

Add Damped Shifted Force (DSF) pairwise electrostatic summation to
nvalchemi-toolkit-ops, including Warp kernels, PyTorch bindings,
benchmarks, accuracy tests, and documentation. Also fixes a
pre-existing bug in the Coulomb matrix energy kernels.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
      functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

### Bug fix: Coulomb matrix energy prefactor

The energy-only matrix kernels (`_coulomb_energy_matrix_kernel` and
`_batch_coulomb_energy_matrix_kernel`) were missing the `0.5`
prefactor that the energy+forces kernels already applied. With a full
(symmetric) neighbor list each pair is visited twice, so the factor
of 0.5 is required to avoid double-counting. Added regression tests
(`TestCoulombMatrixEnergyInvariance`) verifying energy consistency
between the two code paths.

### DSF Warp kernels

Four NVIDIA Warp kernel launchers for DSF electrostatics with
O(N) scaling:

- `dsf` / `dsf_pbc` -- CSR neighbor list format (non-periodic /
  periodic)
- `dsf_matrix` / `dsf_matrix_pbc` -- dense neighbor matrix format

Each kernel computes per-atom energies, analytical forces, self-energy
corrections, and (for PBC variants) the virial stress tensor. Supports
float32 and float64 precision, single-system and batched modes.

### DSF PyTorch bindings

`dsf_coulomb()` provides a unified PyTorch interface that accepts
both CSR and dense neighbor matrix formats and dispatches to the
appropriate Warp kernel variant. Returns `(energy,)`,
`(energy, forces)`, or `(energy, forces, virial)` depending on flags.

Charge gradients are supported via a straight-through estimator for
use with geometry-dependent charges in MLIP training.

### Performance benchmarks

The existing `benchmark_electrostatics.py` is extended with DSF
support. Benchmarks both CSR (list) and dense matrix neighbor
formats to catch performance regressions in either path. Also
adjusted `max_neighbors` estimation for better memory utilization
(previous heuristic was ~50x overestimate).

- **3 new backends**: `nvalchemiops_dsf` (Warp, matrix format),
  `nvalchemiops_dsf_csr` (Warp, CSR format), `torch_dsf` (pure
  PyTorch reference)
- **Refactored neighbor construction**: extracted `build_neighbors()`
  for per-run format switching (CSR vs matrix), replacing previously
  inlined construction
- **New CLI options**: `--method {ewald,pme,dsf,both,all}`,
  `--neighbor-format {list,matrix,both}`, `--dtype {float32,float64}`
- **Extended scaling range**: single systems up to 46^3 supercell
  (~195k atoms); batched systems up to 12,288 x 16 atoms

**Results (RTX 6000 Ada, float32, single system, energy+forces):**

| Atoms | DSF | PME | Ewald |
|------:|----:|----:|------:|
| 54k | 1.2 ms | 94 ms | 257 ms |
| 128k | 2.5 ms | 304 ms (matrix only; list OOM) | OOM |
| 195k | 3.7 ms | OOM | OOM |

DSF is ~80x faster than PME and ~210x faster than Ewald at 54k
atoms. DSF scales to 195k atoms in under 4 ms where both Ewald
and PME run out of 48 GB VRAM.

**Memory (54k atoms):** DSF ~200 MB vs PME ~9.3 GB vs Ewald ~17 GB.

**float32 vs float64 (DSF, 128k atoms):** 2.5 ms vs 6.7 ms
(~2.6x slower in float64).

### Accuracy benchmarks

New standalone `benchmark_dsf_accuracy.py` comparing DSF against
OpenMM PME reference:

- **Systems**: (1) NaCl rock salt crystal with random displacements
  (supercell sizes 6/8/11, up to ~10k atoms); (2) 0.1M NaCl in TIP3P
  water (~9k atoms, built with OpenMM/amber14)
- **Parameter sweep**: cutoff {9, 12, 15} A x alpha {0.1, 0.2, 0.3} x
  {float32, float64}
- **Reference**: OpenMM PME with `ewald_error_tolerance = 1e-6` (pure
  electrostatic system, no LJ or bonded exclusions)
- **Metrics per (cutoff, alpha, dtype)**:
  - Energy absolute error (eV) and per-atom error (meV/atom)
  - Energy-difference RMSE across 16 configurations (meV/atom) --
    measures MLIP-relevant accuracy
  - Force RMS and max error (eV/A) vs reference norms
- **Output**: CSV files named
  `dsf_accuracy_<system>_<gpu_sku>.csv`

**Representative results (NaCl in water, ~26.9k atoms, vs OpenMM
PME):**

| alpha | R_c (A) | Energy err (meV/atom) | Force RMS err (eV/A) |
|------:|--------:|----------------------:|---------------------:|
| 0.2 | 12 | ~1.66 | ~0.027 (~0.6% of ref) |
| 0.2 | 15 | ~0.10 | ~0.027 |
| 0.3 | 9 | ~0.08 | ~0.051 (~1.1% of ref) |

**float32 vs float64 accuracy:** essentially identical -- force
error differences appear only in the 4th-5th decimal place. This
confirms float32 is safe for MLIP-scale DSF calculations.

### Documentation

- **User guide** (`docs/userguide/components/electrostatics.md`):
  DSF section with motivation, mathematical background, usage examples
  (basic, PBC, matrix, charge gradients, batched, alpha=0), parameter
  guidance, method comparison table, applicability/limitations,
  references. Per-method autograd capability table. PBC consistency
  fix in basic example.
- **API references**: `docs/modules/torch/electrostatics.rst` (DSF
  Coulomb section), `docs/modules/warp/electrostatics.rst` (DSF
  Kernels section)
- **Benchmark docs**: `docs/benchmarks/electrostatics.md` updated
  with DSF section and CLI options
- **Landing pages**: `docs/index.md` and `README.md` updated to
  list DSF in key capabilities

### Minor cleanups

- Reformat multi-line tuple in `spline.py`
- Add `openmm>=8.0.0` to `benchmark-requires.txt`
- Remove 3 stale PME entries from electrostatics `__all__`

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage
      expectations?

**DSF test suites** (4,135 test lines total):

- `test/interactions/electrostatics/test_dsf.py` -- Warp kernel tests:
  energy/force correctness, CSR vs matrix consistency, PBC, batched,
  charge gradients, alpha=0, dtype parametrization
- `test/interactions/electrostatics/bindings/torch/test_dsf.py` --
  PyTorch binding tests: end-to-end correctness, autograd,
  virial, batched, format dispatch
- `test/interactions/electrostatics/bindings/torch/test_coulomb.py` --
  Coulomb matrix energy invariance regression tests

## Checklist

- [x] I have read and understand the
      [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help
> conduct pull request reviews. We encourage contributors to read and
> consider suggestions made by Greptile, but note that human
> maintainers will provide the necessary reviews for merging:
> Greptile's comments are **not** a qualitative judgement of your code,
> nor is it an indication that the PR will be accepted/rejected. We
> encourage the use of emoji reactions to Greptile comments, depending
> on their usefulness and accuracy.

### CODE REVIEW

File: benchmarks/interactions/electrostatics/accuracy/benchmark_dsf_accuracy.py [hunk 1]
```diff
@@ -0,0 +1,1179 @@
...
+
+
+def load_config(config_path: Path) -> dict:
+    """Load benchmark configuration from YAML file."""
+    with open(config_path) as f:
+        config = yaml.safe_load(f)
+    return config
+
+
+def estimate_max_neighbors(cutoff: float, density: float, safety: float) -> int:
```
Comment (zubatyuk) [inline benchmarks/interactions/electrostatics/accuracy/benchmark_dsf_accuracy.py]: use upstream function instead.

**Comment by zubatyuk** [comment]
Refactored the DSF module to adopt the dynamics launch system pattern,
consolidating PBC/non-PBC into runtime flags and restructuring kernels
by execution mode. Tests expanded from 47 to 275 DSF-specific tests.
Benchmark `torch_dsf [matrix]` KeyError fixed.

## Warp kernel changes (`nvalchemiops/interactions/electrostatics/dsf.py`)

Restructured from 4 PBC-specific kernels to 4 execution-mode kernels:

- `_dsf_csr_single_kernel` / `_dsf_csr_batch_kernel`
- `_dsf_matrix_single_kernel` / `_dsf_matrix_batch_kernel`

Each kernel accepts a `use_pbc: bool` runtime flag. For non-PBC calls,
the launcher creates zero-sized Warp arrays for `cell` and shifts
(no array initialization inside kernels). Energy accumulation uses
`wp.atomic_add` (unchanged).

Replaced 4 manual launchers + 4 overload dicts with 2 public functions
(`dsf_csr`, `dsf_matrix`) using `resolve_execution_mode` + `launch_family`
from the dynamics launch system. Old `dsf`/`dsf_pbc`/`dsf_matrix_pbc`
names removed (clean break).

## PyTorch binding changes (`nvalchemiops/torch/.../dsf.py`)

Consolidated 4 custom ops into 2 (`_dsf_csr_op`, `_dsf_matrix_op`) with
PBC as optional parameters. `dsf_coulomb()` dispatch simplified from
4-way to 2-way branch. `scatter_add_` retained for charge gradient
corrections (segmented_sum incompatible with PyTorch autograd graph).

## Export changes (`nvalchemiops/interactions/electrostatics/__init__.py`)

Updated: `dsf`, `dsf_pbc`, `dsf_matrix`, `dsf_matrix_pbc` replaced with
`dsf_csr`, `dsf_matrix`.

## Benchmark fix (`benchmark_electrostatics.py`)

Fixed `run_torch_dsf()` KeyError when `neighbor_format="matrix"`. The
function now supports both CSR and matrix formats by converting matrix
to COO on-the-fly for the reference implementation.

## Tests

Warp tests (`test/interactions/electrostatics/test_dsf.py`):

- Migrated all 47 existing tests to new API (`dsf_csr`/`dsf_matrix`)
- Added 7 new test classes: matrix batched, batch PBC, matrix float32,
  matrix charge grad, batch virial, matrix alpha=0, matrix force
  conservation
- Total: 119 tests (CPU + GPU)

PyTorch tests (`test/.../bindings/torch/test_dsf.py`):

- Added 7 new test classes: batch+PBC, matrix autograd, batch autograd,
  matrix float32, PBC autograd, matrix torch.compile, batch virial
- Total: 156 tests (CPU + GPU)

## Verification

```text
5,217 passed, 66 skipped, 0 failed
```

Benchmarks: 324 configurations (162 torch + 162 torch_dsf), all pass.
Performance unchanged from pre-refactoring baseline (RTX 6000 Ada,
float32).
